### PR TITLE
Use DataFusionError instead of ArrowError in SendableRecordBatchStream

### LINKED ISSUE #5039: Use DataFusionError in SendableRecordBatchStream
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The is a significant amount of logic converting from DataFusionError to ArrowError and back, a large amount of which originates from the fact `SendableRecordBatchStream` returns `ArrowError` instead of `DataFusionError`

**Describe the solution you'd like**

It seems strange that the DataFusion stream abstraction would not return DataFusion's error type. I would like to change `SendableRecordBatchStream` to 

```
pub trait RecordBatchStream: Stream<Item = Result<RecordBatch, DataFusionError>> {
    /// Returns the schema of this `RecordBatchStream`.
    ///
    /// Implementation of this trait should guarantee that all `RecordBatch`'s returned by this
    /// stream should have the same schema as returned from this method.
    fn schema(&self) -> SchemaRef;
}
```

**Describe alternatives you've considered**
We could not do this

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #5039.

# Rationale for this change
See #5039 

# What changes are included in this PR?
Replace ArrowError with DataFusionError in DF context

# Are these changes tested?
Yes

# Are there any user-facing changes?
User will get `DatafusionError::ArrowError` instead `arrow::ArrowError` for datafusion contexts

### CODE REVIEW

**Comment by comphead** [comment]
@tustvold please check whenever you have time, the code rotting fast :)

File: datafusion/core/src/physical_plan/joins/nested_loop_join.rs [hunk 1]
```diff
@@ -400,12 +399,9 @@ impl NestedLoopJoinStream {
                     let mut left_indices_builder = UInt64Builder::new();
                     let mut right_indices_builder = UInt32Builder::new();
                     let left_right_indices = match indices_result {
-                        Err(_) => {
-                            // TODO why the type of result stream is `Result<T, ArrowError>`, and not the `DataFusionError`
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/nested_loop_join.rs:404]: :laughing:

File: datafusion/core/src/physical_plan/joins/sort_merge_join.rs [hunk 2]
```diff
@@ -1032,7 +1030,8 @@ impl SMJStream {
                 .columns()
                 .iter()
                 .map(|column| take(column, &streamed_indices, None))
-                .collect::<ArrowResult<Vec<_>>>()?;
+                .collect::<Result<Vec<_>, ArrowError>>()
+                .map_err(Into::<DataFusionError>::into)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/sort_merge_join.rs:1034]:
```suggestion
                .collect::<Result<Vec<_>, ArrowError>>()?;
```

File: datafusion/core/src/physical_plan/joins/sort_merge_join.rs [hunk 3]
```diff
@@ -1045,7 +1044,8 @@ impl SMJStream {
                         .columns()
                         .iter()
                         .map(|column| take(column, &buffered_indices, None))
-                        .collect::<ArrowResult<Vec<_>>>()?
+                        .collect::<Result<Vec<_>, ArrowError>>()
+                        .map_err(Into::<DataFusionError>::into)?
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/sort_merge_join.rs:1048]:
```suggestion
                        .collect::<Result<Vec<_>, ArrowError>>()?
```

File: datafusion/core/src/physical_plan/joins/utils.rs [hunk 4]
```diff
@@ -821,7 +820,7 @@ pub(crate) fn build_batch_from_indices(
         };
         columns.push(array);
     }
-    RecordBatch::try_new(Arc::new(schema.clone()), columns)
+    RecordBatch::try_new(Arc::new(schema.clone()), columns).map_err(Into::into)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/utils.rs:823]:
```suggestion
    Ok(RecordBatch::try_new(Arc::new(schema.clone()), columns)?)
```

File: datafusion/core/src/physical_plan/file_format/file_stream.rs [hunk 5]
```diff
@@ -260,7 +258,11 @@ impl<F: FileOpener> FileStream<F> {
                         self.file_stream_metrics.time_scanning_until_data.stop();
                         self.file_stream_metrics.time_scanning_total.stop();
                         let result = result
-                            .and_then(|b| self.pc_projector.project(b, partition_values))
+                            .and_then(|b| {
+                                self.pc_projector
+                                    .project(b, partition_values)
+                                    .map_err(|e| ArrowError::ExternalError(e.into()))
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/file_stream.rs:264]: Is this needed, `project` already returns `ArrowError` AFAICT?

File: datafusion/core/src/physical_plan/filter.rs [hunk 6]
```diff
@@ -239,20 +238,21 @@ struct FilterExecStream {
...
+) -> Result<RecordBatch> {
     predicate
         .evaluate(batch)
         .map(|v| v.into_array(batch.num_rows()))
-        .map_err(DataFusionError::into)
         .and_then(|array| {
             Ok(as_boolean_array(&array)?)
                 // apply filter array to record batch
-                .and_then(|filter_array| filter_record_batch(batch, filter_array))
+                .and_then(|filter_array| {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/filter.rs:248]: I think this and_then is redundant as it is called on an `Ok()`

File: datafusion/core/src/physical_plan/filter.rs [hunk 7]
```diff
@@ -239,20 +238,21 @@ struct FilterExecStream {
...
+                    filter_record_batch(batch, filter_array).map_err(Into::into)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/filter.rs:249]:
```suggestion
                    Ok(filter_record_batch(batch, filter_array)?)
```

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 8]
```diff
@@ -843,8 +839,9 @@ fn sort_batch(
                     }),
                 )
             })
-            .collect::<ArrowResult<Vec<ArrayRef>>>()?,
-    )?;
+            .collect::<Result<Vec<ArrayRef>, ArrowError>>()?,
+    )
+    .map_err(Into::<DataFusionError>::into)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort.rs:844]:
```suggestion
    )?;
```

File: datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs [hunk 9]
```diff
@@ -462,7 +461,7 @@ impl SortedPartitionByBoundedWindowStream {
         if let Some(columns_to_show) = columns_to_show {
             let n_generated = columns_to_show[0].len();
             self.prune_state(n_generated)?;
-            RecordBatch::try_new(schema, columns_to_show)
+            RecordBatch::try_new(schema, columns_to_show).map_err(Into::into)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/windows/bounded_window_agg_exec.rs:464]:
```suggestion
            Ok(RecordBatch::try_new(schema, columns_to_show)?)
```

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs [hunk 10]
```diff
@@ -348,28 +348,26 @@ impl WindowAggStream {
...
+            )?)
         }
         let columns = transpose(partition_results)
             .iter()
             .map(|elems| concat(&elems.iter().map(|x| x.as_ref()).collect::<Vec<_>>()))
             .collect::<Vec<_>>()
             .into_iter()
-            .collect::<ArrowResult<Vec<ArrayRef>>>()?;
+            .collect::<Result<Vec<ArrayRef>, ArrowError>>()
+            .map_err(Into::<DataFusionError>::into)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:362]:
```suggestion
            .collect::<Result<Vec<ArrayRef>, ArrowError>>()?;
```

File: datafusion/core/src/physical_plan/windows/window_agg_exec.rs [hunk 11]
```diff
@@ -348,28 +348,26 @@ impl WindowAggStream {
...
 
         // combine with the original cols
         // note the setup of window aggregates is that they newly calculated window
         // expression results are always appended to the columns
         let mut batch_columns = batch.columns().to_vec();
         // calculate window cols
         batch_columns.extend_from_slice(&columns);
-        RecordBatch::try_new(self.schema.clone(), batch_columns)
+        RecordBatch::try_new(self.schema.clone(), batch_columns).map_err(Into::into)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/windows/window_agg_exec.rs:370]:
```suggestion
        Ok(RecordBatch::try_new(self.schema.clone(), batch_columns)?)
```

**Review by tustvold — APPROVED** [review]
This looks good to me, mostly just some suggestions to use `?` instead of `map_err(Into::into)` in a couple of places.

Given the intrusiveness of this change, I think we should leave this open for a few days in case there are any objections.

@alamb @andygrove

File: datafusion/core/src/physical_plan/file_format/file_stream.rs (hunk 5 above)
Comment (comphead) [reply datafusion/core/src/physical_plan/file_format/file_stream.rs:264]: This is project from DataFusion, it returns DataFusionError

**Comment by comphead** [comment]
Thanks @tustvold for your review. Fixed all the comments

**Comment by jackwener** [comment]
cc @liukun4515 , You may also be interested in this.

**Review by alamb — APPROVED** [review]

**Pull request merged**